---
title: "P_infestans_CNV_workflow"
author: "Brian J. Knaus"
date: "12/5/2019"
output:
  html_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


---


This project included many files that were organized into many directories.
This is an attempt to document the project workflow throughout the directories.


## *.gvcf.gz processing


```
proc_gvcf/proc_gvcf.R
```

This script inputs files from the directory `gvcf/` as a starting point of the project.
This directory contained over 25 GB of data.
GitHub [presently recommends repositories be kept under 1 GB](https://help.github.com/en/github/managing-large-files/what-is-my-disk-quota) (accessed 2019-12-05).
So you will have to recreate these files.
In order to recreate these you will have to access the sequences from the SRA, map the to the T30-4 reference, and recreate the gvcf files.


This script processes the `.gvcf` files to infer peaks of allele balance.
Among the outputs are two files (per sample) written to `./peaks/` that include the estimate of the peaks and the counts of heterozygous positions in each window.


## Peak processing


```
peak_proc/peak_cat.R
```

This file processes the peaks from `../proc_gvcf`.
It also needs as input `pinf_ref/pinf_ref_wins.csv`.
The resulting output includes: pinf_freq_wins.csv, and pinf_cnts_wins.csv


## Genic CNV

```
gene_to_ploid/gene_to_ploid.R
```

This script takes our processed peak information and uses it to call a copy number for each gene.
Output includes the file `gene_ploidy.csv`.
This is a sample by gene matrix containing a copy number call for each gene in each sample.


## CNV VCF

```
cnv_vcf/cnv_vcf.Rmd
```

This script combines several forms of data into one "VCF-like" file.
Some data was published by other groups during our project so we combine this data with our previous data.
The result is the file "pitg_cnv.vcf.gz".
This is not a valid file under the VCF specification.
But we'd been working with a lot of VCF data and this seemed like a reasonable way to organize the data.


## Distance from expectation

```
cnv_vcf/dfeab.Rmd
```

As a non-parametric form of confidence in our CNV calls we added the distance from expectation.
This script reads in "pitg_cnv.vcf.gz" and outputs "pitg_cnv_abpdfe.vcf.gz".


## AARD

```
aard_ploid/aard_dips.Rmd
aard_ploid/aard_trips.Rmd
```

These scripts use (G/C) adjusted average read depth (AARD) to set a base ploid for individuals that appear primarily diploid or triploid.
They create the files "pitg_cnv_abpdfe_2n.vcf.gz" and "pitg_cnv_abpdfe_3n.vcf.gz".


## Ploidy calls to one file

```
aard_ploid/combine_ploides.Rmd
```

This script combines the ploidy files into one file.
It creates the file "pitg_cnv_```Sys.Date()```.vcf.gz".



