---
title: "Geo_ploid"
author: "Brian J. Knaus"
date: "February 15, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = 'center')
knitr::opts_chunk$set(fig.width = 12)
knitr::opts_chunk$set(fig.height = 6)
```




```{r, results='hide'}
library(vcfR)
#vcf <- read.vcfR("../cnv_vcf/pitg_cnv.vcf.gz")
#vcf <- read.vcfR("../aard_ploid/pitg_cnv_20170327.vcf.gz")
#vcf <- read.vcfR("../aard_ploid/pitg_cnv_20170404.vcf.gz")
#vcf <- read.vcfR("../aard_ploid/pitg_cnv_20170518.vcf.gz", checkFile = FALSE)
#
vcf <- read.vcfR("../aard_ploid/pitg_cnv_20191206.vcf.gz", checkFile = FALSE)
```


```{r}
vcf
queryMETA(vcf)
```



## Filter on AARD


```{r}
vcf
aard <- extract.gt(vcf, element = "AARD", as.numeric = TRUE)
myAard <- colMeans(aard)
#vcf@gt <- vcf@gt[,c(TRUE, myAard >= 12)]
#vcf  <- vcf[,sample = names(myAard[myAard >= 12])]
vcf  <- vcf[,sample = names(myAard[myAard >= 12 & !is.na(myAard)])]
vcf
#colnames(vcf@gt)
```


## ABPDFE

```{r}
ABPDFE <- extract.gt(vcf, element = "ABPDFE", as.numeric = TRUE)
is.na(vcf@gt[,-1][ ABPDFE > 0.9 & !is.na(ABPDFE) ]) <- TRUE
```


## Count gene ploidies


```{r}
abp <- extract.gt(vcf, element = "ABP", as.numeric = TRUE)

pmat <- matrix(ncol=ncol(abp), nrow = 5)
colnames(pmat) <- colnames(abp)

#rownames(pmat) <- c("5","4","3","2","NA")
rownames(pmat) <- c("NA","2","3","4","5")

pmat["5",] <- colSums(abp == 5, na.rm = TRUE)
pmat["4",] <- colSums(abp == 4, na.rm = TRUE)
pmat["3",] <- colSums(abp == 3, na.rm = TRUE)
pmat["2",] <- colSums(abp == 2, na.rm = TRUE)
pmat["NA",] <- apply(abp, MARGIN = 2, function(x){ sum(is.na(x)) })
```


## Omit NA

```{r}
pmat <- pmat[-1,]
```


## Filter taxa


Omit non-infestans

```{r}
pmat <- pmat[,grep("F18", colnames(pmat), invert = TRUE), drop = FALSE]
pipo <- pmat[,grep("00Ip5|Ipom_1.2|Ipom_2.4|Pipo5|P3001", colnames(pmat), invert = FALSE)]
pmat <- pmat[,grep("00Ip5|Ipom_1.2|Ipom_2.4|Pipo5|P3001", colnames(pmat), invert = TRUE)]
pmir <- pmat[,grep("00M_410|CBS.678.85|PIC99167|Pmir5|PIC99114|P7722", colnames(pmat), invert = FALSE)]
pmat <- pmat[,grep("00M_410|CBS.678.85|PIC99167|Pmir5|PIC99114|P7722", colnames(pmat), invert = TRUE)]
ppar <- pmat[,grep("CJ01A1|INRA.310|P10297|P1569|P1976", colnames(pmat), invert = FALSE)]
pmat <- pmat[,grep("CJ01A1|INRA.310|P10297|P1569|P1976", colnames(pmat), invert = TRUE)]
pand <-  pmat[,grep("PaX|EC3425|EC3394|P13803", colnames(pmat), invert = FALSE)]
pmat <-  pmat[,grep("PaX|EC3425|EC3394|P13803", colnames(pmat), invert = TRUE)]
```


Omit 1306

```{r}
pmat <- pmat[,grep("1306", colnames(pmat), invert = TRUE)]
```


Fix some names


```{r}
#colnames(pmat)
colnames(pmat)[grep("06[[:punct:]]3928A", colnames(pmat))] <- "06_3928A"
colnames(pmat)[grep("M[[:punct:]]0182896", colnames(pmat))] <- "M-0182896"
colnames(pmat)[grep("FP[[:punct:]]GCC", colnames(pmat))] <- "FP-GCC"
```


## Write to file


```{r, eval=FALSE}
#write.table(pmat, file = "pmat.csv", sep = ",", col.names = TRUE, row.names = TRUE)
```




## Subset to lineages


```{r}
colnames(pmat)
herb1 <- pmat[,grep("^M-0182896$|^Pi1889$|^P8140$|^P13346$", colnames(pmat)), drop = FALSE]
us1 <- pmat[,grep("^Kew126$|^Kew122$|^DDR7602$|^P8141$|^P6515$|^LBUS$|^P8844$", colnames(pmat)), drop = FALSE]
us8 <- pmat[,grep("^RS2009P1$|^US0400009$", colnames(pmat)), drop = FALSE]
us11 <- pmat[,grep("^FP-GCC$", colnames(pmat)), drop = FALSE]
us18 <- pmat[,grep("^P10127$", colnames(pmat)), drop = FALSE]
us22 <- pmat[,grep("^US100006$|^P17777$|^IN2009T1$", colnames(pmat)), drop = FALSE]
us23 <- pmat[,grep("^BL2009P4$|^FL2009P4$", colnames(pmat)), drop = FALSE]
us24 <- pmat[,grep("^ND822Pi$", colnames(pmat)), drop = FALSE]
#h1306 <- pmat[,grep("^H1306_1$|^H1306_2$|^H1306_3$", colnames(pmat)), drop = FALSE]
mx <- pmat[,grep("^P3685$|^P6634$|^P7036$|^P3683$|^P8143$|^P6636$|^P8144$|^PIC97207$|^PIC97605$|^PIC97630$|^PIC97146$|^PIC97335$|^PIC97785$|^PIC97750$|^P10650$|^PIC98372", colnames(pmat)), drop = FALSE]
sa <- pmat[,grep("^PCZ026$|^PCZ098$|^PCO038$|^PCZ033$|^P13198$|^P13527$|^P13626$", colnames(pmat)), drop = FALSE]
eu <- pmat[,grep("^P6570$|^06_3928A$|^NL07434$", colnames(pmat)), drop = FALSE]
```


## Sort


```{r}
herb1 <- herb1[,sort.int(herb1['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
us1   <- us1[,sort.int(us1['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
us8   <- us8[,sort.int(us8['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
us22  <- us22[,sort.int(us22['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
us23  <- us23[,sort.int(us23['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]

mx <- mx[,sort.int(mx['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
sa <- sa[,sort.int(sa['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
eu <- eu[,sort.int(eu['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
```

## Create a sorted matrix

```{r}
pmat2 <- pmat
pmat2 <- pmat2[,sort.int(pmat2['2',] * 100, decreasing = TRUE, index.return = TRUE)$ix]
```


## Convert to frequencies

```{r}
# Omit NA
#pmat <- pmat[-1,]
#pmat2 <- pmat2[-1,]

pmat <- sweep(pmat, MARGIN = 2, STATS = colSums(pmat), FUN = "/")
pmat2 <- sweep(pmat2, MARGIN = 2, STATS = colSums(pmat2), FUN = "/")

herb1 <- sweep(herb1, MARGIN = 2, STATS = colSums(herb1), FUN = "/")
us1 <- sweep(us1, MARGIN = 2, STATS = colSums(us1), FUN = "/")
us8 <- sweep(us8, MARGIN = 2, STATS = colSums(us8), FUN = "/")
us11 <- sweep(us11, MARGIN = 2, STATS = colSums(us11), FUN = "/")
us18 <- sweep(us18, MARGIN = 2, STATS = colSums(us18), FUN = "/")
us22 <- sweep(us22, MARGIN = 2, STATS = colSums(us22), FUN = "/")
us23 <- sweep(us23, MARGIN = 2, STATS = colSums(us23), FUN = "/")
us24 <- sweep(us24, MARGIN = 2, STATS = colSums(us24), FUN = "/")

mx <- sweep(mx, MARGIN = 2, STATS = colSums(mx), FUN = "/")
sa <- sweep(sa, MARGIN = 2, STATS = colSums(sa), FUN = "/")
eu <- sweep(eu, MARGIN = 2, STATS = colSums(eu), FUN = "/")

#h1306 <- sweep(h1306, MARGIN = 2, STATS = colSums(h1306), FUN = "/")

pipo <- sweep(pipo, MARGIN = 2, STATS = colSums(pipo), FUN = "/")
pmir <- sweep(pmir, MARGIN = 2, STATS = colSums(pmir), FUN = "/")
ppar <- sweep(ppar, MARGIN = 2, STATS = colSums(ppar), FUN = "/")
pand <- sweep(pand, MARGIN = 2, STATS = colSums(pand), FUN = "/")

```



## Plot


```{r}
library(RColorBrewer)
```


```{r}
palette(brewer.pal(n=5, name = "Dark2"))

barplot(herb1, space = 0, las = 3, col=1:5)
title(main = "HERB-1")
title(ylab = "Genes")

barplot(us1, space = 0, las = 3, col=1:5)
title(main = "US-1")
title(ylab = "Genes")

barplot(us8, space = 0, las = 3, col=1:5)
title(main = "US-8")
title(ylab = "Genes")

barplot(us11, space = 0, las = 3, col=1:5)
title(main = "US-11")
title(ylab = "Genes")

barplot(us18, space = 0, las = 3, col=1:5)
title(main = "US-18")
title(ylab = "Genes")

barplot(us22, space = 0, las = 3, col=1:5)
title(main = "US-22")
title(ylab = "Genes")

barplot(us23, space = 0, las = 3, col=1:5)
title(main = "US-23")
title(ylab = "Genes")

barplot(us24, space = 0, las = 3, col=1:5)
title(main = "US-24")
title(ylab = "Genes")

#barplot(h1306, space = 0, las = 3, col=1:5)
#title(main = "H1306")
#title(ylab = "Genes")

barplot(mx, space = 0, las = 3, col=1:5)
title(main = "Mexico")
title(ylab = "Genes")

barplot(sa, space = 0, las = 3, col=1:5)
title(main = "S. America")
title(ylab = "Genes")

barplot(eu, space = 0, las = 3, col=1:5)
title(main = "Europe")
title(ylab = "Genes")

```




```{r}
library(RColorBrewer)
palette(brewer.pal(n=5, name = "Dark2"))

#myMat <- matrix(ncol=8, nrow=3)
#myMat[1,] <- 1
#myMat[2,] <- 2:9
#myMat[3,] <- rep(10:11, each = 4)

myMat <- matrix(ncol=22, nrow=3)
myMat[1,] <- 1
myMat[2,1:4] <- 2
myMat[2,5:10] <- 3
myMat[2,11:12] <- 4
myMat[2,13] <- 5
myMat[2,14] <- 6
myMat[2,15:17] <- 7
myMat[2,18:19] <- 8
myMat[2,20] <- 9
myMat[3,1:13] <- 10
myMat[3,14:18] <- 11
myMat[3,19:20] <- 12

myMat[,21:ncol(myMat)] <- 13

#png("ploidy_bp.png", width = 6.5, height = 6.5, units = "in", res = 300)
#pdf("ploidy_bp.pdf", width = 6.5, height = 6.5)

layout( mat = myMat, widths = c(rep(1, times=21), 1.1) )

par(oma = c(0,4,0,0))
par(mar=c(5,0,2,0))

barplot(pmat2, space = 0, las = 2, col=1:5, yaxt = "n")
title(main = "All samples", cex.main = 0.9)
#title(ylab = "Genes")

axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)

par(mar=c(5,0.1,2,0.1))

barplot(herb1, space = 0, las = 2, col=1:5, yaxt = "n")
title(main = "HERB-1", line = 0.4, cex.main = 0.9)

axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)

barplot(us1, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-1", line = 0.4, cex.main = 0.9)

barplot(us8, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-8", line = 0.4, cex.main = 0.9)

barplot(us11, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-11", line = 0.4, cex.main = 0.9)

barplot(us18, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-18", line = 0.4, cex.main = 0.9)

barplot(us22, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-22", line = 0.4, cex.main = 0.9)

barplot(us23, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-23", line = 0.4, cex.main = 0.9)

barplot(us24, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "US-24", line = 0.4, cex.main = 0.9)

barplot(mx, space = 0, las = 2, col=1:5, yaxt = "n")
title(main = "Mexico", line = 0.4, cex.main = 0.9)

axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)

barplot(sa, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "S. America", line = 0.4, cex.main = 0.9)

barplot(eu, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "Europe", line = 0.4, cex.main = 0.9)

#title(ylab = "Genes (%)", outer = TRUE)
title(ylab = "Genes", outer = TRUE)

plot(c(0,2), c(0,2), type = "n", frame.plot = FALSE, axes = FALSE, xlab = "", ylab = "")
legend(-0.8, 1.25, legend = c("5x", "4x", "3x", "2x"), fill = 4:1, bty = 'n', cex = 2, x.intersp = 0.1)
#legend(-0.8, 1.25, legend = c(">4x", "4x", "3x", "2x"), fill = 4:1, bty = 'n', cex = 2, x.intersp = 0.1)

#dev.off()

par(mfrow=c(1,1))
palette('default')

```


## Plots for map



```{r}
palette(brewer.pal(n=5, name = "Dark2"))
par(mar = c(0,0,0,0))

#png("us_ploid.png", width = 400, height = 400, bg = "transparent")
barplot(cbind(us8, us11, us18, us22, us23, us24), space = 0, las = 3, col=1:5, yaxt = "n", names.arg = rep("", times=10))
#dev.off()

#png("mx_ploid.png", width = 400, height = 400, bg = "transparent")
barplot(mx, space = 0, col=1:5, yaxt = "n", names.arg = rep("", times=ncol(mx)))
#dev.off()

#png("sa_ploid.png", width = 400, height = 400, bg = "transparent")
barplot(sa, space = 0, col=1:5, yaxt = "n", names.arg = rep("", times=ncol(sa)))
#dev.off()

#png("eu_ploid.png", width = 400, height = 400, bg = "transparent")
barplot(eu, space = 0, col=1:5, yaxt = "n", names.arg = rep("", times=ncol(eu)))
#dev.off()

```


## Non Pinf


```{r}

library(RColorBrewer)
palette(brewer.pal(n=5, name = "Dark2"))

par(mar=c(6,4,4,2))

barplot(pmir, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "P. mirabilis")
axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)
barplot(pipo, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "P. ipomoeae")
axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)
barplot(ppar, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "P. parasitica")
axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)
barplot(pand, space = 0, las = 3, col=1:5, yaxt = "n")
title(main = "P. andina")
axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)
```



## FC


```{r}
fc <- structure(list(Samples = structure(c(4L, 5L, 6L, 7L, 8L, 9L, 
12L, 13L, 2L, 3L, 11L, 1L, 10L), .Label = c("CBS.678.85", "FL2009P4", 
"Ipom_1.2", "PIC97136", "PIC97146", "PIC97335", "PIC97442", "PIC97750", 
"PIC97785", "PIC99114", "Pipo5", "US0400009", "US100006"), class = "factor"), 
   Average = c(479.4754652, 393.7044766, 438.9400884, 483.8872061, 
   479.313672, 399.0762326, 575.9737849, 471.6404267, 746.3585317, 
   426.3907895, 449.3997873, 438.9027461, 511.608482), SD = c(0, 
   19.15846798, 8.028315706, 0, 12.09914441, 0, 8.293913352, 
   15.1386842, 20.73050066, 6.559255165, 175.1055099, 28.49094471, 
   13.07142118), CV = c(0, 0.048662053, 0.018290231, 0, 0.025242644, 
   0, 0.01439981, 0.032097936, 0.027775526, 0.015383201, 0.389643064, 
   0.064914027, 0.025549657)), .Names = c("Samples", "Average", 
"SD", "CV"), class = "data.frame", row.names = c(NA, -13L))
```



```{r}

myRegex <- paste(as.character(fc$Samples), collapse = "$|^")
myRegex <- paste("^", myRegex, "$", sep = "")

pmat3 <- pmat2[,grep(myRegex, colnames(pmat2))]

myRegex <- paste(colnames(pmat3), collapse = "$|^")
myRegex <- paste("^", myRegex, "$", sep = "")

fc3 <- fc[grep(myRegex, fc$Samples),]

pmat3 <- pmat3[,as.character(fc3$Samples)]

#cbind(colnames(pmat3), as.character(fc3$Samples))



#par(mfrow=c(2,1))
#png("fc_ab.png", width = 450, height = 450)

layout(matrix(1:2, ncol=1), heights = c(1.4,2))

par(mar=c(1,4,0,2))
barplot(fc3$Average, space = 0, las = 2, col="#99d8c9", yaxt = "n", ylim = c(0,800))
abline(h=480)
abline(h=722)
axis(side = 2, at=seq(100,700, by=100), las = 2)
title(ylab="Genome size (MBp)", line = 3)

par(mar=c(6,4,0,2))
barplot(pmat3, space = 0, las = 2, col=1:5, yaxt = "n")
axis(side = 2, at = seq(0,1,by=0.2), labels = c("0%","20%","40%","60%","80%","100%"),las = 2)
title(ylab="Genes")

#dev.off()


par(mar=c(5,4,4,2))

```


## 1 row


```{r}
#pdf("cline_bp.pdf", width = 5.5, height = 2.5, pointsize = 8)
par(oma = c(0,0,0,0))
par(mar=c(6,3,2,0))
par(mgp = c(2,0.2,0))
barplot(pmat2, space = 0, las = 2, col=1:5, yaxt = "n")
axis(side = 2, at = seq(0, 1, by = 0.2), line = -1, las = 2, 
     labels = c("0%", "20%", "40%", "60%", "80%", "100%"))
title(ylab = "Genes", line = 2)
par(mar=c(5,4,4,2))
#dev.off()
```


```{r}
myMat <- matrix(ncol=47, nrow=1)
myMat[1, 1:16] <- 1
myMat[1, 17:23] <- 2
myMat[1, 24:27] <- 3
myMat[1, 28:34] <- 4
myMat[1, 35:36] <- 5
myMat[1, 37:37] <- 6
myMat[1, 38:38] <- 7
myMat[1, 39:41] <- 8
myMat[1, 42:43] <- 9
myMat[1, 44:44] <- 10
myMat[1, 45:47] <- 11


#pdf("geo_bp.pdf", width = 5.5, height = 2.5, pointsize = 10)

layout( mat = myMat)
par(oma = c(0, 3, 0, 0))
par(mar = c(6,0.1,1,0.1))
par(mgp = c(2,0.2,0))
barplot(mx, space = 0, las = 2, col=1:5, yaxt = "n")
axis(side = 2, at = seq(0, 1, by = 0.2), line = -1, las = 2, 
     labels = c("0%", "20%", "40%", "60%", "80%", "100%"))
title(ylab = "Genes", line = 2, outer = TRUE)
barplot(sa, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(herb1, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us1, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us8, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us11, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us18, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us22, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us23, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(us24, space = 0, las = 2, col=1:5, yaxt = "n")
barplot(eu, space = 0, las = 2, col=1:5, yaxt = "n")

#dev.off()


par(mar = c(5,4,4,2))
```


